doc/reference/index.rst

```diff
@@ -23,3 +23,265 @@ API Reference
    usb/index.rst
    usermode/index.rst
    runtime_conf/index.rst
+
+Terminology
+***********
+
+The following terms may be used as shorthand API tags to indicate the
+allowed calling context (thread, ISR, pre-kernel), the effect of a call
+on the current thread state, and other behavioral characteristics.
+
+:ref:`api_term_reschedule`
+   if executing the function reaches a reschedule point
+:ref:`api_term_sleep`
+   if executing the function can cause the invoking thread to sleep
+:ref:`api_term_no-wait`
+   if a parameter to the function can prevent the invoking thread from
+   trying to sleep
+:ref:`api_term_isr-ok`
+   if the function can always be safely called from interrupt context
+   even if it may return an error in that case
+:ref:`api_term_pre-kernel-ok`
+   if the function can be safely called before the kernel has been fully
+   initialized, even if it may return an error in that case
```

Comment:
What's the actual use case here?  This seems overspecific to my ears.  The "PRE_KERNEL" init states are note themselves well defined.  Can we not write this as "before the kernel launches application threads and enters main()"?

Reply:
The use case is #21090 where the API to request a clock must be usable both in interrupt kernel and before the kernel starts.  I'm using **pre-kernel-ok** because the way to detect the case and avoid doing bad things is by checking `k_is_pre_kernel()` which verifies the condition you identify.